modules/gapi/src/executor/gapi_tbb_executor.cpp

```diff
@@ -0,0 +1,362 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2020 Intel Corporation
+
+#include "gapi_tbb_executor.hpp"
+
+#if defined(HAVE_TBB)
+#include "gapi_itt.hpp"
+
+#include <opencv2/gapi/own/assert.hpp>
+#include "logger.hpp" // GAPI_LOG
+
+#include <tbb/task.h>
+#include <tbb/task_arena.h>
+#include <memory> //unique_ptr
+
+#include <atomic>
+#include <condition_variable>
+
+#include <chrono>
+
+#define ASSERT(expr)          GAPI_Assert(expr)
+
+#define LOG_INFO(tag, ...)    GAPI_LOG_INFO(tag, __VA_ARGS__)
+#define LOG_WARNING(tag, ...) GAPI_LOG_WARNING(tag, __VA_ARGS__)
+#define LOG_DEBUG(tag, ...)   GAPI_LOG_DEBUG(tag, __VA_ARGS__)
+
+
+#ifdef OPENCV_WITH_ITT
+const __itt_domain* cv::gimpl::parallel::gapi_itt_domain = __itt_domain_create("GAPI Context");
+#endif
+
+namespace cv{ namespace gimpl { namespace parallel {
+
+inline void assert_graph_is_running(tbb::task* root)
+{
+   //tbb::task::wait_for_all block calling thread until task ref_count is dropped to 1
+   //So if the root task ref_count is greater than 1 graph still has a job to do and
+   //according wait_for_all() has not yet returned
+   ASSERT(root->ref_count() > 1);
+}
+
+enum class use_tbb_scheduler_bypass {
+   yes,
+   no
+};
+
+enum class wake_tbb_master {
+   yes,
+   no
+};
+
+//made template to break circular dependencies
+template<typename body_t>
+struct functor_task : tbb::task {
+   body_t body;
+
+   template<typename arg_t>
+   functor_task(arg_t&& a) : body(std::forward<arg_t>(a)){}
+
+   tbb::task * execute() override
+   {
+      assert_graph_is_running(parent());
+
+      auto reuse_current_task = body();
+      //if needed, say TBB to execute current task once again
+      return (use_tbb_scheduler_bypass::yes ==  reuse_current_task) ? (recycle_as_continuation(), this) : nullptr;
+   }
+   ~functor_task(){
+      assert_graph_is_running(parent());
```

Comment:
Be careful with exceptions in destructors.

Details: https://en.cppreference.com/w/cpp/error/uncaught_exception

Reply:
thanks for the warning, but this intentionally is not a exception, but and `assert` either `CV_` one or a hand-crafted one

Reply:
Changed to be enabled in debug only